crates/ruff/src/importer.rs

```diff
@@ -0,0 +1,314 @@
+                if level.map_or(true, |level| level == 0) {
+                    if let Some(module) = module {
+                        self.import_from_map.insert(module.as_str(), import);
+                    }
+                }
+            }
+            _ => {
+                unreachable!("Expected StmtKind::Import | StmtKind::ImportFrom");
+            }
+        }
+
+        // Store a reference to the last top-level import statement.
+        self.trailing_import = Some(import);
+    }
+
+    /// Add an import statement to import the given module.
+    ///
+    /// If there are no existing imports, the new import will be added at the top
+    /// of the file. Otherwise, it will be added after the most recent top-level
+    /// import statement.
+    pub fn add_import(&self, import: &AnyImport) -> Edit {
+        let required_import = import.to_string();
+        if let Some(stmt) = self.trailing_import {
+            // Insert after the last top-level import.
+            let (prefix, location, suffix) =
+                end_of_statement_insertion(stmt, self.locator, self.stylist);
+            let content = format!("{prefix}{required_import}{suffix}");
+            Edit::insertion(content, location)
+        } else {
+            // Insert at the top of the file.
+            let (prefix, location, suffix) =
+                top_of_file_insertion(self.python_ast, self.locator, self.stylist);
+            let content = format!("{prefix}{required_import}{suffix}");
+            Edit::insertion(content, location)
+        }
+    }
+
+    /// Return the top-level [`Stmt`] that imports the given module using `StmtKind::ImportFrom`.
+    /// if it exists.
+    pub fn get_import_from(&self, module: &str) -> Option<&Stmt> {
+        self.import_from_map.get(module).copied()
+    }
+
+    /// Add the given member to an existing `StmtKind::ImportFrom` statement.
+    pub fn add_member(&self, stmt: &Stmt, member: &str) -> Result<Edit> {
+        let mut tree = match_module(self.locator.slice(stmt))?;
+        let import_from = match_import_from(&mut tree)?;
+        let aliases = match_aliases(import_from)?;
+        aliases.push(ImportAlias {
+            name: NameOrAttribute::N(Box::new(Name {
+                value: member,
+                lpar: vec![],
+                rpar: vec![],
+            })),
+            asname: None,
+            comma: aliases.last().and_then(|alias| alias.comma.clone()),
+        });
+        let mut state = CodegenState {
+            default_newline: self.stylist.line_ending(),
+            default_indent: self.stylist.indentation(),
+            ..CodegenState::default()
+        };
+        tree.codegen(&mut state);
+        Ok(Edit::replacement(
+            state.to_string(),
+            stmt.location,
+            stmt.end_location.unwrap(),
+        ))
+    }
+}
+
+/// Find the end of the last docstring.
+fn match_docstring_end(body: &[Stmt]) -> Option<Location> {
```

Comment:
GitHub isn't picking it up, but everything from here down is moved from `rules/isort/rules/add_required_import.rs` (which now uses this `Importer` struct).